framework/include/utils/RankFourTensor.h

```diff
@@ -693,3 +705,54 @@ RankFourTensorTempl<T>::fillSymmetric21FromInputVector(const T2 & input)
   (*this)(2, 0, 1, 0) = input[19];
   (*this)(1, 0, 2, 0) = input[19];
 }
+
+template <typename T>
+template <typename T2,
+          typename std::enable_if<(RankFourTensorTempl<T2>::N2 * sizeof(T2) >
+                                   EIGEN_STACK_ALLOCATION_LIMIT),
+                                  int>::type>
+RankFourTensorTempl<T>
+RankFourTensorTempl<T>::inverse() const
+{
+  // Allocate on the heap if you're going to exceed the stack size limit
+
+  // The inverse of a 3x3x3x3 in the C_ijkl*A_klmn = de_im de_jn sense is
+  // simply the inverse of the 9x9 matrix of the tensor entries.
+  // So all we need to do is inverse _vals (with the appropriate row-major
+  // storage)
+
+  RankFourTensorTempl<T> result;
+  Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> mat(9, 9);
+  for (auto i : make_range(9 * 9))
+    mat(i) = _vals[i];
+
+  mat = mat.inverse();
+
+  for (auto i : make_range(9 * 9))
+    result._vals[i] = mat(i);
+
+  return result;
+}
+
+template <typename T>
+template <typename T2,
+          typename std::enable_if<(RankFourTensorTempl<T2>::N2 * sizeof(T2) <=
```

Comment:
I think N2 in this class evaluates to 9, but you need `9*9` which is the size of the big tensor entry matrix. I don't know if you've seen my Mandel notation PR, but maybe that would be a path for reducing the computational cost here. 

At first I was surprised to see a tests timing out due to such a change, but the heap allocated dynamic matrix inverse routine in Eigen cannot make certain optimizations, such as loop unrolling, making it a bit slower.

With the symmetric r4 tensor, all 36 entries fit on the stack for our default ADReal size, that has the twofold benefit of fewer heap allocations/deallocations, and a substantially more optimized inversion routine.

Reply:
Ya, NEML maintains separate symmetric and full tensor classes like what you're doing (and uses Mandel notation too!).

I warn you though that you're going down a dark path in doing this -- we (me/crystal plasticity people) also need skew rank two tensors so then if you're being complete you end up with weird RankFour tensors that are SkewSymmetric (skew in the first 2 indices, symmetric in the second 2), SymmetricSkew, and SkewSkew in addition to the "SymmetricSymmetric" type you're implementing now.  Of course you always fall back on general RankFours, but I got overexcited in NEML.

You're correct on `N2`, I want `N4`.  I'll fix it, but let this round of checks go first to make sure I don't have anything else to fix.

Reply:
I wonder if the skew can be dealt with through clever templating :-/. Just to make sure I got this. The skew for an index pair means swapping indices changes the sign?

Reply:
I suppose the real challenge is figuring out the symmetries of the return types of various operations. Yikes.